src/group_theory/exponent.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2021 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import group_theory.order_of_element
+import algebra.punit_instances
+import algebra.gcd_monoid.finset
+import ring_theory.int.basic
+
+/-!
+# Exponent of a group
+
+This file defines the exponent of a group, or more generally a monoid. For a group `G` it is defined
+to be the minimal `n≥1` such that `g ^ n = 1` for all `g ∈ G`. For a finite group `G` it is equal to
+the lowest common multiple of the order of all elements of the group `G`.
+
+## Main definitions
+
+* `monoid.exponent_exists` is a predicate on a monoid `G` saying that there is some positive `n`
+  such that `g ^ n = 1` for all `g ∈ G`.
+* `monoid.exponent` defines the exponent of a monoid `G` as the minimal positive `n` such that
+  `g ^ n = 1` for all `g ∈ G`, by convention it is `0` if no such `n` exists.
+* `add_monoid.exponent_exists` the additive version of `monoid.exponent_exists`.
+* `add_monoid.exponent` the additive version of `monoid.exponent`.
+
+## Main results
+
+* `lcm_order_eq_exponent`: For a finite group `G`, the exponent is equal to the `lcm` of the order
+  of its elements.
+
+## TODO
+* Compute the exponent of cyclic groups.
+* Refactor the characteristic of a ring to be the exponent of its underlying additive group.
+-/
+
+universe u
+
+variables (G : Type u) [monoid G]
+
+open_locale classical
+
+namespace monoid
+
+/-A predicate on a monoid saying that there is a positive integer `n` such that `g ^ n = 1`
+  for all `g`.-/
+@[to_additive]
+def exponent_exists  := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
+
+/-The exponent of a group is the smallest positive integer `n` such that `g ^ n = 1` for all `g ∈ G`
+if it exists, otherwise it is zero by convention.-/
+@[to_additive]
+noncomputable def exponent :=
+if h : exponent_exists G then nat.find h else 0
+
+@[to_additive exponent_nsmul_eq_zero]
+lemma pow_exponent_eq_one (g : G) : g ^ exponent G = 1 :=
+begin
+  by_cases exponent_exists G,
+  { simp_rw [exponent, dif_pos h],
+    exact (nat.find_spec h).2 g},
+  { simp_rw [exponent, dif_neg h, pow_zero] }
+end
+
+@[to_additive nsmul_eq_mod_exponent]
+lemma pow_eq_mod_exponent {n : ℕ} (g : G): g ^ n = g ^ (n % exponent G) :=
+calc g ^ n = g ^ (n % exponent G + exponent G * (n / exponent G)) : by rw [nat.mod_add_div]
+  ... = g ^ (n % exponent G) : by simp [pow_add, pow_mul, pow_exponent_eq_one]
+
+@[to_additive]
+lemma exponent_pos_of_exists (n : ℕ) (hpos : 0 < n)
+(hG : ∀ g : G, g ^ n = 1) : 0 < exponent G :=
+begin
+  have h : ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1,
+      { exact ⟨n, hpos, hG⟩ },
```

Comment:
```suggestion
  have h : ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1 := ⟨n, hpos, hG⟩,
```

or is this a line length issue? regardless I don't think the original indenting is mathlib's style